### PR TITLE
Implement interpret with AWS Comprehend

### PR DESCRIPTION
**Pull request by raphkim**
*Issue #, if available:*

*Description of changes:*
- Implement `AWSPredictionsPlugin.interpret()` using AWS Comprehend
- Implement Predictions configuration logic for `interpret`
- Add unit tests for configuration
- Add instrumentation tests

**Note:** instrumentation test is blocked by a bug that will be fixed by #345 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: testutils/src/main/java/com/amplifyframework/testutils/sync/SynchronousPredictions.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A utility to perform synchronous calls to the {@link PredictionsCategory}.
+ * This code is not well suited for production use, but is useful in test
+ * code, where we want to make a series of sequential assertions after
+ * performing various operations.
+ */
+public final class SynchronousPredictions {
+    private static final long PREDICTIONS_OPERATION_TIMEOUT_MS = TimeUnit.SECONDS.toMillis(60);
```
[the remark this answers is not in the document]
Comment (raphkim) [reply testutils/src/main/java/com/amplifyframework/testutils/sync/SynchronousPredictions.java:36]: unfortunately, a single `interpret` call can make up to 5 API calls...

Meanwhile, 1 minute was definitely not my intent LOL. Looks like I forgot to change it after debugging.

File: aws-predictions/src/test/resources/configuration-with-interpret.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+  "defaultRegion": "us-east-1",
+  "interpret": {
+    "interpretText": {
+      "region": "us-west-2",
+      "type": "SENTIMENT",
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-predictions/src/test/resources/configuration-with-interpret.json:6]: This is based off of the actual `amplifyconfiguration.json` file that was generated by CLI. I will add support for lower case also.

File: aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/AWSPredictionsInterpretTest.java [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+    /**
+     * Assert that interpret fails for blank input.
+     * AWS Comprehension will not accept a blank input, so
+     * an exception is thrown.
+     * @throws Exception if prediction fails
+     */
+    @Test(expected = PredictionsException.class)
+    public void testInterpretFailsForNullInput() throws Exception {
+        predictions.interpret("", InterpretOptions.defaultInstance());
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/AWSPredictionsInterpretTest.java:67]: `default` vs `defaults`? Thoughts?

File: aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/AWSPredictionsInterpretTest.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        InterpretResult result = predictions.interpret(
+                Assets.readAsString("sample-text-en.txt"),
+                InterpretOptions.defaultInstance()
+        );
+        assertNotNull(result);
+
+        // Assert detected key phrases are not empty
+        List<KeyPhrase> keyPhrases = result.getKeyPhrases();
+        assertNotNull(keyPhrases);
+        assertFalse(keyPhrases.isEmpty());
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/AWSPredictionsInterpretTest.java:162]: I would love to, but currently I lack a way to compare a list of features. I'm wondering if I should make two feature instances evaluate `.equals()` to `true` if the value matches without looking at the confidence score.

File: aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/AWSPredictionsInterpretTest.java [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    }
+
+    /**
+     * Assert that interpret correctly labels syntax.
+     * @throws Exception if prediction fails
+     */
+    @Test
+    @SuppressWarnings("MagicNumber") // TODO: What's a better way to compare results?
+    public void testSyntaxDetection() throws Exception {
+        final String sampleText = "I am inevitable.";
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/AWSPredictionsInterpretTest.java:191]: My original intent was to actually bring in large paragraphs, but I ended up using something short for debugging purposes and never changed it back. I may do just that.

File: aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/TestConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        if (singleton == null) {
+            singleton = new TestConfiguration();
+        }
+        return singleton;
+    }
+
+    private static void configureAmplify(Context context) throws AmplifyException {
+        // Obtain Amplify Configuration
+        final int configId = Resources.getRawResourceId(context, AMPLIFY_CONFIGURATION_IDENTIFIER);
+        AmplifyConfiguration configuration = AmplifyConfiguration.fromConfigFile(context, configId);
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/TestConfiguration.java:65]: I have this here so that a customer can try running the instrumentation test with multiple versions of `amplifyconfiguration.json`. All they need to do to test using a different configuration would be to change the name of constant variable inside `TestConfiguration`. In fact, I may actually require this feature myself to test different types of interpret configuration since CLI officially supports one interpret resource per project.

File: aws-predictions/src/main/java/com/amplifyframework/predictions/aws/AWSPredictionsPluginConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-predictions/src/main/java/com/amplifyframework/predictions/aws/AWSPredictionsPluginConfiguration.java:2]: darn

File: aws-predictions/src/main/java/com/amplifyframework/predictions/aws/AWSPredictionsPluginConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+     * Constructs an instance of {@link AWSPredictionsPluginConfiguration} from
+     * the plugin configuration JSON object.
+     * @param configurationJson the plugin configuration
+     * @return the configuration object for AWS Predictions Plugin
+     * @throws PredictionsException if configuration is missing or malformed
+     */
+    @NonNull
+    public static AWSPredictionsPluginConfiguration fromJson(JSONObject configurationJson) throws PredictionsException {
+        if (configurationJson == null) {
+            throw new PredictionsException(
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-predictions/src/main/java/com/amplifyframework/predictions/aws/AWSPredictionsPluginConfiguration.java:56]: sounds a bit excessive, since it's doing the same thing with one extra step without any benefits

File: aws-predictions/src/main/java/com/amplifyframework/predictions/aws/adapter/EntityTypeAdapter.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    @SuppressWarnings("checkstyle:all") private EntityTypeAdapter() {}
+
+    /**
+     * Converts the entity type string returned by AWS Comprehend
+     * into a format supported by Amplify Predictions.
+     * @param entity Entity type returned by AWS Comprehend
+     * @return Amplify's {@link EntityType} enum
+     */
+    @NonNull
+    public static EntityType fromComprehend(@NonNull String entity) {
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-predictions/src/main/java/com/amplifyframework/predictions/aws/adapter/EntityTypeAdapter.java:39]: that sounds excessive, considering that there is no performance benefits to doing that and doesn't make it more readable necessarily.

File: aws-predictions/src/main/java/com/amplifyframework/predictions/aws/service/AWSComprehendService.java [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+        try {
+            // First obtain the dominant language to begin analysis
+            final Language dominantLanguage = fetchPredominantLanguage(text);
+            final LanguageType language = dominantLanguage.getValue();
+
+            // Actually analyze text in the context of dominant language
+            final Sentiment sentiment = fetchSentiment(text, language);
+            final List<KeyPhrase> keyPhrases = fetchKeyPhrases(text, language);
+            final List<Entity> entities = fetchEntities(text, language);
+            final List<Syntax> syntax = fetchSyntax(text, language);
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-predictions/src/main/java/com/amplifyframework/predictions/aws/service/AWSComprehendService.java:96]: That... is actually exactly what I should have been going for haha

**Comment by raphkim** [comment]
oops not yet ready for review...

File: aws-predictions/src/test/resources/configuration-with-interpret.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-predictions/src/test/resources/configuration-with-interpret.json:6]: on second thought, I think it is unnecessary, since we don't want to encourage them to directly edit these anyways. The best way to edit these without error would be to update via: 
```
amplify update predictions
amplify push
```

File: aws-predictions/src/main/java/com/amplifyframework/predictions/aws/configuration/AWSInterpretConfiguration.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+     * Construct an instance of {@link AWSInterpretConfiguration} from
+     * plugin configuration JSON object.
+     * @param configurationJson the plugin configuration
+     * @return the interpret configuration
+     * @throws JSONException if interpret configuration is malformed
+     */
+    @NonNull
+    public static AWSInterpretConfiguration fromJson(@NonNull JSONObject configurationJson) throws JSONException {
+        // Use default values if the section is missing
+        if (!configurationJson.has("interpret")) {
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-predictions/src/main/java/com/amplifyframework/predictions/aws/configuration/AWSInterpretConfiguration.java:54]: I think having non-standard key snuck in there for no-op comment purpose is fine.

File: aws-predictions/src/main/java/com/amplifyframework/predictions/aws/AWSPredictionsPluginConfiguration.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-predictions/src/main/java/com/amplifyframework/predictions/aws/AWSPredictionsPluginConfiguration.java:56]: I will require it for now, since predictions resources need to be added via CLI anyways

File: aws-predictions/src/main/java/com/amplifyframework/predictions/aws/service/AWSComprehendService.java [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+
+            onSuccess.accept(InterpretResult.builder()
+                    .language(dominantLanguage)
+                    .sentiment(sentiment)
+                    .keyPhrases(keyPhrases)
+                    .entities(entities)
+                    .syntax(syntax)
+                    .build());
+        } catch (PredictionsException exception) {
+            onError.accept(exception);
```
Comment (royjit) [inline aws-predictions/src/main/java/com/amplifyframework/predictions/aws/service/AWSComprehendService.java:106]: Are we failing the whole operation if any of the above service fails? I think iOS sends partial result back to the user.

File: aws-predictions/src/main/java/com/amplifyframework/predictions/aws/service/AWSComprehendService.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-predictions/src/main/java/com/amplifyframework/predictions/aws/service/AWSComprehendService.java:96]: on second thought, I think I'll revisit this in the future. This may require each component to be made into its own predictions service method, and I want to see how the final Predictions service with all of the APIs implemented looks before I do that.

File: aws-predictions/src/main/java/com/amplifyframework/predictions/aws/service/AWSComprehendService.java [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+                    serviceException,
+                    "See attached service exception for more details."
+            );
+        }
+
+        // Convert AWS Comprehend's detection result to Amplify-compatible format
+        List<KeyPhrase> keyPhrases = new ArrayList<>();
+        for (com.amazonaws.services.comprehend.model.KeyPhrase comprehendKeyPhrase : result.getKeyPhrases()) {
+            TargetText target = new TargetText(
+                    comprehendKeyPhrase.getText(),
```
Comment (royjit) [inline aws-predictions/src/main/java/com/amplifyframework/predictions/aws/service/AWSComprehendService.java:234]: When using beginOffset and text, make sure that text containing special characters are counted correctly. For example text with smileys. "face \uD83D\uDE10" whats the length of this string? 17 or 7

Comment (royjit) [reply aws-predictions/src/main/java/com/amplifyframework/predictions/aws/service/AWSComprehendService.java:234]: Applies to all places that uses beginOffset and length for text.

File: aws-predictions/src/main/java/com/amplifyframework/predictions/aws/adapter/EntityTypeAdapter.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (jamesonwilliams) [reply aws-predictions/src/main/java/com/amplifyframework/predictions/aws/adapter/EntityTypeAdapter.java:39]: Yea, it's a little subjective I guess. But I appreciate your reasoning. Sounds good!

File: aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/AWSPredictionsInterpretTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jamesonwilliams) [reply aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/AWSPredictionsInterpretTest.java:162]: Ohh, gotcha. Hm. Well, I guess you could eval against a list of expected `KeyPhrase`s themselves?
```
assertEquals(
    Arrays.asList(
        KeyPhrase.builder().value("the").confidence(30).build(),
        KeyPhrase.builder().value("key").confidence(20).build(),
        KeyPhrase.builder().value("phrases").confidence(50).build()
    ),
    result.getPhrases()
);
```
Or if the `value` returns are not stable/predictable, and you'd rather not do literal assertions on them, maybe just check their ordering by confidence:
```
assertEquals(
    Arrays.asList("the", "key", "phrases"),
    Observable.fromIterable(results.getKeyPhrases())
        .map(KeyPhrase::getValue)
        .toSortedList() // This actually already will sort on confidence, as is, right? Cause of compareTo in Feature?
        .blockingGet()
);
```

Ah crap ... just realized in re-reading it, but we are missing `equals`/`hashCode()`/`toString()` on some of the `KeyPhrase`, `TextFeature`, and `Feature`?

File: aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/TestConfiguration.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jamesonwilliams) [reply aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/TestConfiguration.java:64]: Ah, gotcha, hm, that makes sense.

To test multiple configs, we'll either have to build some teardown in to the `Amplify` facade, or not use it, in tests. (This way, we're not "stuck" with whatever the first instance of `new PredictionsCategory()` happens to have, in that `static` field inside of `Amplify`.) We could just rework our tests to go against the _`Category`_, directly. This would make our tests much more of real-life "integration" tests.

In other words, like:
```
PredictionsCategory predictions = new PredictionsCategory();
predictions.addPlugin(new AWSPredictionsPlugin());
CategoryConfiguration predictionsConfiguration =
    AmplifyConfiguration.fromConfigFile(context, Resources.getRawResourceId(".../path/..."))
        .forCategoryType(CategoryType.PREDICTIONS);
predictions.configure(predictionsConfiguration, context);
predictions.initialize(context);
```

File: aws-predictions/src/main/java/com/amplifyframework/predictions/aws/service/AWSComprehendService.java (hunk 12 above)
Comment (raphkim) [reply aws-predictions/src/main/java/com/amplifyframework/predictions/aws/service/AWSComprehendService.java:106]: discussed. JS also fails the entire operation so this will stay, but I can definitely see the merit in returning partial result. Perhaps we can do something similar to what we have in API category? `InterpretResult` can also contain error field, which will contain the list of service errors encountered. The error callback will only trigger on network error or validation error

File: aws-predictions/src/main/java/com/amplifyframework/predictions/aws/service/AWSComprehendService.java (hunk 13 above)
Comment (raphkim) [reply aws-predictions/src/main/java/com/amplifyframework/predictions/aws/service/AWSComprehendService.java:234]: Java appears to handle it appropriately actually (length 7). I will leave it as is for now.

File: aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/AWSPredictionsInterpretTest.java [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+
+    private static SynchronousPredictions predictions;
+
+    /**
+     * Configure Amplify and Predictions plugin before the tests.
+     * @throws Exception if configuration fails
+     */
+    @BeforeClass
+    public static void setUpOnce() throws Exception {
+        TestConfiguration.configureIfNotConfigured();
```
Comment (jamesonwilliams) [inline aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/AWSPredictionsInterpretTest.java:55]: Can we update this to use "the new way"? OR want to do it in a different PR?

File: testutils/src/main/java/com/amplifyframework/testutils/sync/SynchronousPredictions.java [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+        this.predictions = predictions;
+    }
+
+    /**
+     * Gets an instance of the Synchronous Predictions utility that
+     * delegates tasks to Amplify.Predictions.
+     * @return new instance of Synchronous Predictions
+     */
+    @NonNull
+    public static synchronized SynchronousPredictions delegatingToAmplify() {
```
Comment (jamesonwilliams) [inline testutils/src/main/java/com/amplifyframework/testutils/sync/SynchronousPredictions.java:51]: I ended up removing these versions of the factory in the other synchronous utils. I want to make it inconvenient for tests to call `Amplify`, actually (since it will pollute state.) If we require that an `PredictionsCategoryBehavior` be supplied in order to constrcut `SyncrhonousPredictions`, then the caller will be guided to use the associated `TestPredictionsCategory.create()` (or whatever.)

File: aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/TestConfiguration.java [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+import com.amplifyframework.testutils.Resources;
+import com.amplifyframework.testutils.sync.SynchronousMobileClient;
+
+import com.amazonaws.mobile.config.AWSConfiguration;
+
+/**
+ * This is a class to help configure Amplify and prepare other test resources
+ * for Predictions integration tests.
+ */
+final class TestConfiguration {
```
Comment (jamesonwilliams) [inline aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/TestConfiguration.java:34]: Can we move over to `TestPredictionsCategory` factory style?

File: aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/AWSPredictionsInterpretTest.java [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+public final class AWSPredictionsInterpretTest {
+
+    private static SynchronousPredictions predictions;
+
+    /**
+     * Configure Amplify and Predictions plugin before the tests.
+     * @throws Exception if configuration fails
+     */
+    @BeforeClass
+    public static void setUpOnce() throws Exception {
```
Comment (jamesonwilliams) [inline aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/AWSPredictionsInterpretTest.java:54]: Note: if you get this to a fresh `PredictionsCategory` each time, then you can make it as `@Before public void setup()`, which will help guarantee fresh state for each individual test.

(Once we get rid of the process singletons, we should kill all of the `@BeforeClass`. Yea, they might be a little more performance than doing setup before each individual test, but at the cost of potential cross-pollution of the individual tests.)

File: aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/TestConfiguration.java (hunk 16 above)
Comment (raphkim) [reply aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/TestConfiguration.java:34]: Yes! I think your refactored logic is way cleaner and makes more sense.

File: aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/AWSPredictionsInterpretTest.java (hunk 14 above)
Comment (raphkim) [reply aws-predictions/src/androidTest/java/com/amplifyframework/predictions/aws/AWSPredictionsInterpretTest.java:55]: I'll follow up in the next PR
